In [3]:
import numpy as np
import pandas as pd
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

**2. Given the following confusion matrix, evaluate (by hand) the model's performance.**
positive: dogs

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

**- In the context of this problem, what is a false positive?**. 
predict more people like dogs than cats but it is false

**- In the context of this problem, what is a false negative?**
predicted people don't like more dogs, but really they like more dogs

**- How would you describe this model?**
- TP: predicted it is dog and   it is true
- FP: predicted  it is dog, but  it is  cat
- FN: predicted  it is dog, but  it is  cat
- TN: predicted it is cat ,  and it is true 


In [4]:
#positive: dogs
TP =  46 # predicted it is dog and   it is true
FP = 13 # predicted  it is dog, but  it is  cat
FN = 7  # predicted  it is dog, but  it is  cat
TN = 34 # predicted it is cat ,  and it is true 

In [6]:
TP, FP, FN, TN

(46, 13, 7, 34)

In [8]:
#accuracy
accuracy = (TP + TN) / (TP + TN + FN + FP)
accuracy

0.8

In [9]:
#recall
recall = TP / (TP + FN)
recall

0.8679245283018868

In [10]:
precision = TP / (TP + FP)
precision

0.7796610169491526

this model is high in Recall rate : 86.79%, 


********************************************

*****************review  by Madeleine*********************

In [71]:
#positive = cat
#create a df
df = pd.DataFrame({'pred_dog': [46, 13], 'pred_cat': [7, 34]}, index=['actual_dog', 'actual_cat'])
df

pred_dog  pred_cat
actual_dog        46         7
actual_cat        13        34

In [72]:
rubric_df = pd.DataFrame([['True Negative', 'False positive'], ['False Negative', 'True Positive']], columns=df.columns, index=df.index)

In [73]:
rubric_df

pred_dog        pred_cat
actual_dog   True Negative  False positive
actual_cat  False Negative   True Positive

In [74]:
# add df and  rubric_df
rubric_df + ': ' + df.values.astype(str)

pred_dog           pred_cat
actual_dog   True Negative: 46  False positive: 7
actual_cat  False Negative: 13  True Positive: 34

In [76]:
# Recall: TP / (TP + FN)
recall = 34 / (34 + 13)
recall

0.723404255319149

In [77]:
# Precision: TP / (TP + FP)
precision = 34 / (34 + 7)
precision

0.8292682926829268

In [78]:
# Accuracy : (TP + TN) / (TP + TN + FP + FN)
accuracy = (34 + 46) / df.values.sum()
accuracy

0.8

*********************************

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

- A. An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. 
Which evaluation metric would be appropriate here? Recall
Which model would be the best fit for this use case?  Model 3

-  positive = Defect
-  negative = No Defect
- we want a recall because we want to catch the actually positive cases

In [4]:
# acquire data
c3 = pd.read_csv('C3.csv')

In [5]:
c3.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [6]:
c3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


**p.cross is giving me the same info as confusion_matrix**

In [ ]:
# model1

In [7]:
pd.crosstab(c3.model1, c3.actual)

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [8]:

confusion_matrix(c3.actual, c3.model1,
                 labels = ('No Defect', 'Defect'))

array([[182,   2],
       [  8,   8]])

In [9]:
#model 2
pd.crosstab(c3.model2, c3.actual)

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [10]:
#model 2
confusion_matrix(c3.actual, c3.model2,
                 labels = ('No Defect', 'Defect'))

array([[103,  81],
       [  7,   9]])

In [18]:
#model 3
pd.crosstab(c3.model3, c3.actual)

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [11]:
#model 3
confusion_matrix(c3.actual, c3.model3,
                 labels = ('No Defect', 'Defect'))

array([[98, 86],
       [ 3, 13]])

In [ ]:
#baseline


In [25]:
c3.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [26]:
# we are adding a new column with the value of our baseline that we chose : Defect
c3['baseline_prediction'] = 'Defect'

In [27]:
c3.head()

actual     model1  model2     model3 baseline_prediction
0  No Defect  No Defect  Defect  No Defect              Defect
1  No Defect  No Defect  Defect     Defect              Defect
2  No Defect  No Defect  Defect  No Defect              Defect
3  No Defect     Defect  Defect     Defect              Defect
4  No Defect  No Defect  Defect  No Defect              Defect

**We choose Recall  because we want to catch the actually positive cases(defect) so we are going to calculate Recall for each model**

In [13]:
subset = c3[c3.actual == 'Defect']
subset.head()


actual     model1     model2  model3
13  Defect  No Defect     Defect  Defect
30  Defect     Defect  No Defect  Defect
65  Defect     Defect     Defect  Defect
70  Defect     Defect     Defect  Defect
74  Defect  No Defect  No Defect  Defect

**RECALL: compare models to reality and get the mean (from the boolean values)**

In [29]:
# Recall model 1 
(subset.model1 == subset.actual).mean()


0.5

In [30]:
# Recall model 2 

(subset.model2 == subset.actual).mean()


0.5625

In [14]:
# Recall model 3 

(subset.model3 == subset.actual).mean()


0.8125

**We will choose model 3  becuase has the highest Recall :81.25%**

*******************************

- ** b. Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?**

positive = Defect
evaluation metric : Precision because the  false positives are expensive.

In [16]:
#model 1
subset = c3[c3.model1 == 'Defect']
subset

actual  model1     model2     model3
3    No Defect  Defect     Defect     Defect
30      Defect  Defect  No Defect     Defect
62   No Defect  Defect  No Defect  No Defect
65      Defect  Defect     Defect     Defect
70      Defect  Defect     Defect     Defect
135     Defect  Defect  No Defect     Defect
147     Defect  Defect  No Defect     Defect
163     Defect  Defect     Defect     Defect
194     Defect  Defect  No Defect     Defect
196     Defect  Defect  No Defect  No Defect

In [17]:
(subset.model1 == subset.actual).mean()

0.8

In [18]:
#model 2
subset = c3[c3.model2 == 'Defect']
(subset.model2 == subset.actual).mean()

0.1

In [19]:
#model 3
subset = c3[c3.model3 == 'Defect']
(subset.model3 == subset.actual).mean()

0.13131313131313133

 **The  model with the best fit for this use case: Model 1**

*********************************

### 4.You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
- b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
- c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

In [21]:
#acquire data

paws = pd.read_csv('gives_you_paws.csv')


In [22]:
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [23]:
#most common class in actual
paws.actual.max()

'dog'

In [24]:
#adding a new column "baseline"
paws['baseline'] = paws.actual.max()

In [25]:
paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

**a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?**

In [26]:
positive = 'dog'
# accuracy -- overall hit rate
#model1
model_accuracy1 = (paws.model1 == paws.actual).mean()
baseline_accuracy = (paws.baseline == paws.actual).mean()

#model2
model_accuracy2 = (paws.model2 == paws.actual).mean()

#model3
model_accuracy3 = (paws.model3 == paws.actual).mean()

#model4
model4_a = (paws.actual == paws.model4).mean()



In [30]:
print(f'''
positive: {positive}

         | baseline | model1 | model2| model3 |  model4 | 
         | -------- | ------ | ------| ------ |  ------ |      
accuracy | {baseline_accuracy:8.1%} | {model_accuracy1:6.1%} | {model_accuracy2:6.1%}| {model_accuracy3:6.1%} | {model4_a:6.1%}  |
''')


positive: dog

         | baseline | model1 | model2| model3 |  model4 | 
         | -------- | ------ | ------| ------ |  ------ |      
accuracy |    65.1% |  80.7% |  63.0%|  51.0% |  74.3%  |



**model 1 and for are better thatn baseline**

*******************************

**b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?**

- Phase I we want accuracy
- For Phase II? if I working on a team that solely deals with dog pictures. Precision



In [39]:
positive = 'dog'
# Precisson -- overall hit rate

In [40]:
#baseline
subset = paws[paws.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

In [41]:
#model1
subset = paws[paws.model1 == positive]
model1_precision = (subset.model1 == subset.actual).mean()

In [42]:
#model2
subset = paws[paws.model2 == positive]
model2_precision = (subset.model2 == subset.actual).mean()

In [43]:
#model3
subset = paws[paws.model3 == positive]
model3_precision = (subset.model3 == subset.actual).mean()

In [44]:
#model4
subset = paws[paws.model4 == positive]
model4_precision = (subset.model4 == subset.actual).mean()

In [45]:
print(f'''
positive: {positive}

         | baseline | model1 | model2| model3 | model4 |   
         | -------- | ------ | ------| ------ | ------ |      
precision| {baseline_precision:8.1%} | {model1_precision:6.1%} | {model2_precision:6.1%}| {model3_precision:6.1%} |{model4_precision:6.1%} |
''')


positive: dog

         | baseline | model1 | model2| model3 | model4 |   
         | -------- | ------ | ------| ------ | ------ |      
precision|    65.1% |  89.0% |  89.3%|  66.0% | 73.1% |



**Model 2 is the besto fro precision**

*************************************

**c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?**

In [ ]:
#Phase I we want accuracy
# For Phase II? if I working on a team that solely deals with cats pictures. precision

In [46]:
positive = 'cat'
# Precisson -- overall hit rate
#model1
subset = paws[paws.model1 == positive]
model1_precision = (subset.model1 == subset.actual).mean()


In [47]:
subsetb = paws[paws.baseline == positive]
subsetb

Empty DataFrame
Columns: [actual, model1, model2, model3, model4, baseline]
Index: []

In [48]:
baseline_precision = (subsetb.baseline == subsetb.actual).mean()

In [49]:
#model2
subset = paws[paws.model2 == positive]
model2_precision = (subset.model2 == subset.actual).mean()

In [56]:
#model3
subset3 = paws[paws.model3 == positive]
model3_precision = (subset3.model3 == subset3.actual).mean()

In [57]:
#model4
subset = paws[paws.model4 == positive]
model3_precision = (subset.model4 == subset.actual).mean()

In [58]:
print(f'''
positive: {positive}

         | baseline | model1 | model2| model3 | model4 | 
         | -------- | ------ | ------| ------ | ------ |      
precision| {baseline_precision:8.1%} | {model1_precision:6.1%} | {model2_precision:6.1%}| {model3_precision:6.1%} | {model4_precision:6.1%} | 
''')


positive: cat

         | baseline | model1 | model2| model3 | model4 | 
         | -------- | ------ | ------| ------ | ------ |      
precision|     nan% |  69.0% |  48.4%|  80.7% |  73.1% | 



### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

In [59]:
#get the columns
paws.columns

Index(['actual', 'model1', 'model2', 'model3', 'model4', 'baseline'], dtype='object')

In [60]:
#get all the columns expect the first one
paws.columns[1:]

Index(['model1', 'model2', 'model3', 'model4', 'baseline'], dtype='object')

In [61]:
#make a list from the names of columns
list(paws.columns[1:])

['model1', 'model2', 'model3', 'model4', 'baseline']

In [62]:
models= list(paws.columns[1:])


- **sklearn.metrics.accuracy_score**

In [63]:
#just for one model
accuracy_score(paws.actual, paws.model1)

0.8074

In [66]:
for col in models:
    acc = accuracy_score(paws.actual, paws[col])
    print(f'The accuracy for {col} is: {acc:.2%}')
    

The accuracy for model1 is: 80.74%
The accuracy for model2 is: 63.04%
The accuracy for model3 is: 50.96%
The accuracy for model4 is: 74.26%
The accuracy for baseline is: 65.08%


- **sklearn.metrics.precision_score**

In [122]:
#just for one model
precision_score(paws.actual, paws.model1, pos_label = 'dog' )

0.8900238338440586

In [68]:
for col in models:
    p = precision_score(paws.actual, paws[col], pos_label = 'dog' )
    print(f'The precision for {col} is: {p:.2%}' )

The precision for model1 is: 89.00%
The precision for model2 is: 89.32%
The precision for model3 is: 65.99%
The precision for model4 is: 73.12%
The precision for baseline is: 65.08%


- **sklearn.metrics.recall_score**

In [129]:
#just for one model
recall_score(paws.actual, paws.model1, pos_label = 'dog')

0.803318992009834

In [70]:
for col in models:
    r= recall_score(paws.actual, paws[col], pos_label = 'dog')
    print(f'The Recall for {col} is: {r:.2%}' )

The Recall for model1 is: 80.33%
The Recall for model2 is: 49.08%
The Recall for model3 is: 50.86%
The Recall for model4 is: 95.57%
The Recall for baseline is: 100.00%


- **sklearn.metrics.classification_report**

In [ ]:
classification_report(paws.actual, paws.model1)

In [132]:
print(classification_report(paws.actual, paws.model1))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [136]:
for col in models:
    print(' ')
    print('Classification Report for :', col )
    print(classification_report(paws.actual, paws.model1))
    print('________________________________________________________')

 
Classification Report for : model1
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

_________________________________________________
 
Classification Report for : model2
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

_________________________________________________
 
Classification Report for : model3
              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89   